proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,322 @@
+
+## Motivation
+
+CAPN aims at providing control-plane level isolation while sharing physical resources 
+among control planes. There exist various ways of creating isolated control-planes. 
+For example, one can create NCPs through cloud providers' Kubernetes services or has 
+a separate cluster, just running NCPs. However, the most straightforward way is directly 
+running the nested control-plane components on the underlying clusters. In this proposal, 
+we present a way of hosting the nested control-plane on the underlying clusters. By 
+using the proposed mechanism, the user can set up the CAPN environment on any standard 
+Kubernetes cluster without using any third-party platforms or resources.
+
+### Goals
+
+- Define the CRD of each nested component.
+- Define how NEtcd, NKAS and NKCM controllers cooperate to create components of NCP.
+
+### Non-Goals
+
+- Define how NCP controller works.
+- Define how NPKI controller works.
+- Define a mechanism/interface that is portable to other existing controllers, e.g., etcd-operator.
+
+## Proposal
+
+### Bootstrap
+
+The NCP controller will need to create some resources before sub-controllers 
+start creating the real control plane components. 
+
+* **ClusterNamespace** is used to store all resources for the NCP.
+
+* **NKAS service**  is used to visit the NKAS. Since IP/domain 
+name is required when generating TLS certificates and kubeconfigs of NKAS, this 
+service needs to be created in advance.
+
+* **NestedCRs** includes NPKI, NEtcd, NKAS, and NKCM. The NCP controller needs to 
+create all CRs for sub-controllers. Among them, NEtcd, NKAS, and NKCM will be 
+created based on the NCPT referred by the NCP.
+
+After creating the above prerequisites, sub-controllers will start creating the workload of each component.
+
+### Creation
+
+By default, we will use the StatefulSet to create each component. The readiness 
+and liveness probe will be used, and we will mark each component as ready only 
+when the corresponding Statefuleset is ready. As the KAS cannot work without 
+available Etcd and the KCM cannot run without KAS, the three components need to 
+be created in the order of NEtcd, NKAS, and NKCM
+
+Each component CRD contains a Template field that defines the pod template. The 
+NCP controller will generate the pod template based on the corresponding addon 
+object defined in the corresponding NCPT. For the first implementation, we will 
+create a set of templates in this repository. In the future, we plan to support 
+using templates from some well-known projects, e.g., kubeadm.
+
+
+The creating process will include six steps:
+![Control Plane Creating Process](creating-control-plane.png)
+
+1. The NCP controller notices the creation of a new NCP CR.
+2. NCP controller creates ClusterNamespace, NPKI, NEtcd, NKAS, NKCM, and the NKAS service.
+3. NPKI controller generates TLS certificates for NEtcd and NKAS, creates two kubeconfigs, 
```

Comment:
I wonder if we could make every component do their own PKI and use these are fields on the components so we could pass them between CRs via reference

Reply:
we could make the kubeconfig work which should be similar to what KCP does and make that a part of the NCP controller like https://cluster-api.sigs.k8s.io/tasks/certs/using-custom-certificates.html & https://cluster-api.sigs.k8s.io/tasks/certs/generate-kubeconfig.html

this would allow implementers a way to override them

Reply:
Wild idea… what if NestedEtcd, NestedAPIServer and, NestedControllerManager could end up going the route of being implemented like ClusterAddons where you could have the standard implementation but you could supply patches to them https://github.com/kubernetes-sigs/cluster-addons/blob/master/walkthrough.md flipping the idea away from templates to more declarative and easier to customize. This would mean we could implement higher level abstractions like we have with templates later but staying very focused on the Resources themselves.